### PR TITLE
116 bring in mypy

### PR DESCRIPTION
**Pull request by tschm**
Some big refactoring. Introduced types...

### CODE REVIEW

**Comment by phschiele** [comment]
Lets bring in mypy better earlier than later! Do you want to make more changes to the config as stated in the comment or merge as is?
Nice use of `numpy.typing`!

**Comment by tschm** [comment]
This won't work for Python 3.8 and 3.9... I will fix it

**Comment by tschm** [comment]
@phschiele Would be good to merge this...

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -23,3 +23,13 @@ repos:
...
   - id: markdownlint-fix
     args: ["--ignore", "book/**/*.md"]
+
+
+- repo: 'https://github.com/pre-commit/mirrors-mypy'
+  rev: v1.4.1
+  hooks:
+    - id: mypy
+      files: cvx
+      # More work needed...
```
Comment (phschiele) [inline .pre-commit-config.yaml:33]: Can be removed?

File: .pre-commit-config.yaml [hunk 2]
```diff
@@ -23,3 +23,13 @@ repos:
...
+      args: [--strict, --ignore-missing-imports, --explicit-package-bases]
+      #args: [--ignore-missing-imports, --explicit-package-bases]
```
Comment (phschiele) [inline .pre-commit-config.yaml:35]: Can also be removed?

File: cvx/linalg/random.py [hunk 3]
```diff
@@ -1,7 +1,9 @@
...
 import numpy as np
 
+from cvx.linalg.types import Matrix
 
-def rand_cov(n):
+
+def rand_cov(n: int) -> Matrix:
     a = np.random.randn(n, n)
-    return np.transpose(a) @ a
+    return np.array(np.transpose(a) @ (a))
```
Comment (phschiele) [inline cvx/linalg/random.py:9]: That was a matrix before, no?

File: cvx/markowitz/builder.py [hunk 4]
```diff
@@ -26,7 +32,7 @@ class _Problem:
     problem: cp.Problem
     model: Dict[str, Model] = field(default_factory=dict)
 
-    def update(self, **kwargs):
+    def update(self, **kwargs: Matrix) -> _Problem:
```
Comment (phschiele) [inline cvx/markowitz/builder.py:35]: Kwargs are a matrix? I thought usually a dict.

File: cvx/markowitz/risk/sample/sample.py [hunk 5]
```diff
@@ -47,7 +47,7 @@ def estimate(self, variables: Dict[str, cp.Variable]) -> cp.Expression:
             )
         )
 
-    def update(self, **kwargs):
+    def update(self, **kwargs: Matrix) -> None:
```
Comment (phschiele) [inline cvx/markowitz/risk/sample/sample.py:50]: Same as above.

File: cvx/linalg/types.py [hunk 6]
```diff
@@ -0,0 +1,6 @@
+# -*- coding: utf-8 -*-
+import numpy as np
+import numpy.typing as npt
+from typing_extensions import TypeAlias
+
+Matrix: TypeAlias = npt.NDArray[np.float64]
```
Comment (phschiele) [inline cvx/linalg/types.py:6]: Two we need to types files or can we reuse one?

**Review by phschiele — APPROVED** [review]
Had a few comments, let me know what you think!

File: cvx/markowitz/builder.py (hunk 4 above)
Comment (tschm) [reply cvx/markowitz/builder.py:34]: https://adamj.eu/tech/2021/05/11/python-type-hints-args-and-kwargs/

Comment (tschm) [reply cvx/markowitz/builder.py:34]: I didn't know that

File: cvx/linalg/types.py (hunk 6 above)
Comment (tschm) [reply cvx/linalg/types.py:5]: I would try to avoid a dependency in linalg on markowitz

File: cvx/markowitz/risk/sample/sample.py (hunk 5 above)
Comment (tschm) [reply cvx/markowitz/risk/sample/sample.py:49]: https://adamj.eu/tech/2021/05/11/python-type-hints-args-and-kwargs/

File: .pre-commit-config.yaml (hunk 2 above)
Comment (tschm) [reply .pre-commit-config.yaml]: removed

File: .pre-commit-config.yaml (hunk 1 above)
Comment (tschm) [reply .pre-commit-config.yaml]: removed

File: cvx/linalg/random.py (hunk 3 above)
Comment (tschm) [reply cvx/linalg/random.py:8]: I am not sure how to explicitly enforce it's that type. np.transpose(a) @ a gives some sort of Any

Comment (phschiele) [reply cvx/linalg/random.py:8]: I see. We can also override mypy by ignoring individual lines if we know they are doing the right thing by adding `# type:ignore` as a comment.

File: cvx/markowitz/risk/sample/sample.py (hunk 5 above)
Comment (phschiele) [reply cvx/markowitz/risk/sample/sample.py:49]: Interesting, thanks for sharing!

**Review by phschiele — APPROVED** [review]
Thanks for adding the changes, looks good now!

**Pull request merged**